### PR TITLE
feat(api): add Konnect API group

### PR DESCRIPTION
**Pull request by pmalek**
**What this PR does / why we need it**:

- move `api/v1beta1` and `api/v1alpha1` under `api/gateway-operator/`
- add `api/konnect` with `v1alpha1` placeholder

**Which issue this PR fixes**

Part of #442 

**Special notes for your reviewer**:

This is a breaking change in the Go API but not in YAML manifests as API  group name does not change.

We do not treat the Go API currently as part of the user contract so we do not label this change as breaking.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: config/crd/bases/konnect.konghq.com_konnectapiauthconfigurations.yaml [hunk 1]
```diff
@@ -0,0 +1,171 @@
+---
+apiVersion: apiextensions.k8s.io/v1
```
Comment (randmonkey) [inline config/crd/bases/konnect.konghq.com_konnectapiauthconfigurations.yaml]: Looks like no corresponding type definitions for the CRD. Was it generated?

File: config/crd/bases/konnect.konghq.com_konnectcontrolplanes.yaml [hunk 2]
```diff
@@ -0,0 +1,206 @@
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
```
Comment (randmonkey) [inline config/crd/bases/konnect.konghq.com_konnectcontrolplanes.yaml]: Ditto this.

**Review by randmonkey — CHANGES_REQUESTED** [review]
Looks like no go types for Konnect CRDs are defined in this PR but we have manifests for two CRDs. If we do not want include konnect CRDs in this PR, I think we should remove the manifests.

File: config/crd/bases/konnect.konghq.com_konnectapiauthconfigurations.yaml (hunk 1 above)
Comment (pmalek) [reply config/crd/bases/konnect.konghq.com_konnectapiauthconfigurations.yaml]: Good catch. 3c98f5c5e980db9a2b23a0bc031b80c49faded45 removes this.

File: config/crd/bases/konnect.konghq.com_konnectcontrolplanes.yaml (hunk 2 above)
Comment (pmalek) [reply config/crd/bases/konnect.konghq.com_konnectcontrolplanes.yaml]: Good catch. 3c98f5c5e980db9a2b23a0bc031b80c49faded45 removes this.

**Review by randmonkey — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -33,6 +33,14 @@
   plugins distributed as container images.
   [400](https://github.com/Kong/gateway-operator/pull/400)
 
+### Changes
```
Comment (czeslavo) [inline CHANGELOG.md:36]: ~I think we could be explicit and call these changes breaking as they technically are, right?~
Nevermind, I missed "We do not treat the Go API currently as part of the user contract so we do not label this change as breaking." in the description. 😅 

```suggestion
### Breaking changes
```

**Review by czeslavo — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (pmalek) [reply CHANGELOG.md:36]: Yes and no. As stated in the PR description we do not consider Go API as part of the user contract. This is not mentioned anywhere though.

The manifests / YAML API is what we keep up to Kubernetes' API versioning standard: https://kubernetes.io/docs/reference/using-api/#api-versioning which this PR does not change.

I can be persuaded to change this to "breaking" though. My priority here is to give users a good hint about what's coming in the upcoming release. Do you feel putting this in `Changes` section does not fit well in that context?

Comment (czeslavo) [reply CHANGELOG.md:36]: I have no hard opinion on that and leaving it as "changes" works as well. 

I initially thought that if a technically public Go API is changed, we might communicate it as breaking as we never know who relies on that. If we do not want to make the Go API public, we could a) make it internal or b) explicitly state somewhere in the docs we do not guarantee its stability in any way.

**Review by mlavacca — CHANGES_REQUESTED** [review]
I thought we were going to add the new konnect CRDs in kubernetes-configuration rather than here. We agreed on putting off the migration of the existing CRDs to the new repo, but for what concerns the new CRDs, why are we splitting them up between here and kubernetes-config?

**Comment by pmalek** [comment]
> I thought we were going to add the new konnect CRDs in kubernetes-configuration rather than here. We agreed on putting off the migration of the existing CRDs to the new repo, but for what concerns the new CRDs, why are we splitting them up between here and kubernetes-config?

There are 2 "types" of CRDs we're talking about here: 

- Kong configuration CRDs which are already in https://github.com/Kong/kubernetes-configuration/tree/main/api/configuration (`KongConsumer`, `KongConsumerGroup`, `KongPlugin` etc) and those that might still end up there or are being actively worked on (`KongService` or `KongRoute`)
- Top level Konnect entity types like `KonnectControlPlane`. This is the type of CRDs that will land in the newly added API group.

The former make sense in both KGO and KIC context. The latter are only used by the operator and are only used in the Konnect context.

---

From what I understand @mlavacca, you're suggesting to introduce the latter also in https://github.com/Kong/kubernetes-configuration under `konnect` API group, leaving KGO API unchanged. 

I'm wondering if there are any benefits to this approach. I'm curious about your thoughts on this. Given that we'd only use those CRDs in the operator I'm not sure there's much value in putting them outside of this repo. 🤔

File: CHANGELOG.md (hunk 3 above)
Comment (pmalek) [reply CHANGELOG.md:36]: @czeslavo PTAL at #445.

**Comment by mlavacca** [comment]
> > I thought we were going to add the new konnect CRDs in kubernetes-configuration rather than here. We agreed on putting off the migration of the existing CRDs to the new repo, but for what concerns the new CRDs, why are we splitting them up between here and kubernetes-config?
> 
> There are 2 "types" of CRDs we're talking about here:
> 
> * Kong configuration CRDs which are already in https://github.com/Kong/kubernetes-configuration/tree/main/api/configuration (`KongConsumer`, `KongConsumerGroup`, `KongPlugin` etc) and those that might still end up there or are being actively worked on (`KongService` or `KongRoute`)
> * Top level Konnect entity types like `KonnectControlPlane`. This is the type of CRDs that will land in the newly added API group.
> 
> The former make sense in both KGO and KIC context. The latter are only used by the operator and are only used in the Konnect context.
> 
> From what I understand @mlavacca, you're suggesting to introduce the latter also in https://github.com/Kong/kubernetes-configuration under `konnect` API group, leaving KGO API unchanged.
> 
> I'm wondering if there are any benefits to this approach. I'm curious about your thoughts on this. Given that we'd only use those CRDs in the operator I'm not sure there's much value in putting them outside of this repo. 🤔

My point is that since we are making the effort to group the CRDs in a unique place, it could make sense to treat this new place as where "Kong CRDs live". No matters if they are shared among KIC or KGO, or just used by KGO, I think that having a centralized place where our APIs live is a good approach, rather than having them scattered across multiple repositories.

Does it make sense to you?

**Comment by mlavacca** [comment]
One additional point to the discussion is there we will inevitably split the konnect CRDs between k8s-conf and kgo if we choose the approach of putting the `KonnectControlPlanes` here: the kubernetes configuration objects will need the `konnectAPIAuthConfiguration` API as it is used in their status, and to avoid circular dependencies we'll have to put such resources in the same repo. We'd end up having the Kong CRDs scattered across multiple repos, AND the same API group scattered across multiple repositories, which is something I think we really need to stay away from.

**Comment by pmalek** [comment]
> One additional point to the discussion is there we will inevitably split the konnect CRDs between k8s-conf and kgo if we choose the approach of putting the `KonnectControlPlanes` here: the kubernetes configuration objects will need the `konnectAPIAuthConfiguration` API as it is used in their status, and to avoid circular dependencies we'll have to put such resources in the same repo. We'd end up having the Kong CRDs scattered across multiple repos, AND the same API group scattered across multiple repositories, which is something I think we really need to stay away from.

I was almost going to nod on this but then I realized that with the approach proposed by @czeslavo in the design doc (https://docs.google.com/document/d/1TxW2ovqPRm-w1td1e_V0CyaET4lpiEBCJ7hiYMZJqng/edit?disco=AAABS5JK2YM) we could drop the idea of having the `KonnectAuthConfiguration` reference in resources living within the CP because the latter would contain the auth config anyway. This way we don't need the `KonnectAuthConfiguration` type information.

Would you agree @mlavacca?

I guess at this point, it's a matter of where we'd **prefer** this CRDs to live. Let's discuss this on today's epic review.

**Comment by pmalek** [comment]
After a discussion with @mlavacca and @czeslavo we've decided that it will be better to place all new Konnect related types and CRDs into https://github.com/Kong/kubernetes-configuration.

With this, https://github.com/Kong/kubernetes-configuration/pull/12 supersedes this PR.

**Pull request closed without merging**